raiden/utils/gas_reserve.py

```diff
@@ -0,0 +1,123 @@
+from typing import Tuple
+
+from raiden.constants import (
+    GAS_REQUIRED_FOR_OPEN_CHANNEL,
+    GAS_REQUIRED_FOR_SET_TOTAL_DEPOSIT,
+    GAS_REQUIRED_FOR_CLOSE_CHANNEL,
+    GAS_REQUIRED_FOR_SETTLE_CHANNEL,
+    GAS_REQUIRED_FOR_UNLOCK_1_LOCKS,
+)
+from raiden.transfer import views
+
+GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_SETTLE = (
+    GAS_REQUIRED_FOR_UNLOCK_1_LOCKS
+)
+GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_CLOSE = (
+    GAS_REQUIRED_FOR_SETTLE_CHANNEL + GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_SETTLE
+)
+GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_OPEN = (
+    GAS_REQUIRED_FOR_CLOSE_CHANNEL + GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_CLOSE
+)
+GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_COMPLETE = (
+    GAS_REQUIRED_FOR_OPEN_CHANNEL +
+    GAS_REQUIRED_FOR_SET_TOTAL_DEPOSIT +
+    GAS_REQUIRED_FOR_SET_TOTAL_DEPOSIT
+)
+
+GAS_RESERVE_ESTIMATE_SECURITY_FACTOR = 1.1
+
+
+def _get_required_gas_estimate(
+    new_channels: int = 0,
+    opened_channels: int = 0,
+    closing_channels: int = 0,
+    closed_channels: int = 0,
+    settling_channels: int = 0,
+    settled_channels: int = 0,
+) -> int:
+    estimate = 0
+
+    estimate += new_channels * GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_COMPLETE
+    estimate += opened_channels * GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_OPEN
+    # the close transaction might fail, so assume that another might need to be sent
+    estimate += closing_channels * GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_OPEN
+    estimate += closed_channels * GAS_REQUIRED_FOR_CHANNEL_LIFECYCLE_AFTER_CLOSE
+    # the settle transaction might fail, so assume that another might need to be sent
```

Comment:
Why? The transaction may fail for two reasons:

- A bug, and sending the same transaction again will not fix anything
- A race condition with our partner, and **his** transaction succeeded, so we don't need to send a second one.

I don't think we need to send a second transaction on any circumstance

Reply:
Makes sense, fixed.